### ISSUE TITLE
System.Linq.Expressions.Expression.Lambda.Compile() returns null (when using non-nullabale reference types)

### ISSUE THREAD

**Issue by sidney-pauly** [dotnet/runtime#35336]
TLDR: When compiling generated code that calls a function on a non-nullable reference type Lambda.Compile() returns null.  

I had the issue where I called Compile() on a generated lambda and it returned null. The Visual Studio debugger displayed \<Internal Error evaluating expression\>:
![d0f32d4d-f9aa-464e-904e-a702e031c929](https://user-images.githubusercontent.com/44103563/80087728-cdd49280-855b-11ea-8d69-4c335acab4f8.jpg)

The code-generating code was as follows:

```
        public static Func<T, VSerializationOptions, VEntityProperty> GenerateSerializationFunctionNotWorking<T>()
        {
            var WriteResultDictType = typeof(Dictionary<string, VEntityProperty>);
            var SerializeMethod = typeof(VSerializeable).GetMethod(nameof(VSerializeable.Serialize));

            ConstructorInfo entityPropConstructor;
            ParameterExpression entityPropConstructorInput;
            ParameterExpression RESULT_FUNC_PARAM_instance = Expression.Parameter(typeof(T), "instance"); // Input of the result func
            ParameterExpression RESULT_FUNC_PARAM_options = Expression.Parameter(typeof(VSerializationOptions), "options"); // Wheather or not to use a short property name
            ParameterExpression RESULT_FUNC_VAR_result = Expression.Variable(typeof(VEntityProperty), "result");

            var vars = new List<ParameterExpression> { RESULT_FUNC_VAR_result };
            var expressions = new List<Expression>();

            ParameterExpression RESULT_FUNC_VAR_subProp = Expression.Variable(WriteResultDictType, "subProp"); // Used to save sub properties. E. g. when serializing a list or a complex subobject
            vars.Add(RESULT_FUNC_VAR_subProp);

            // RESULT_FUNC_EXPRESSION Serialize the property
            expressions.Add(Expression.Assign(RESULT_FUNC_VAR_subProp, Expression.Call(Expression.Convert(RESULT_FUNC_PARAM_instance, typeof(VSerializeable)), SerializeMethod, new[] { RESULT_FUNC_PARAM_options })));

            // Get the constructor that accepts the results of Serialize.
            entityPropConstructor = typeof(VEntityProperty).GetConstructor(new[] { WriteResultDictType })!;
            entityPropConstructorInput = RESULT_FUNC_VAR_subProp;

            // RESULT_FUNC_EXPRESSION Create a new entity property with the serialization result and assign it to entityProp
            expressions.Add(Expression.Assign(RESULT_FUNC_VAR_result, Expression.New(entityPropConstructor, new Expression[] { Expression.Convert(entityPropConstructorInput, entityPropConstructor.GetParameters().First().ParameterType) })));

            // RESULT_FUNC_EXPRESSION return the result
            expressions.Add(RESULT_FUNC_VAR_result);

            var lamda = Expression.Lambda<Func<T, VSerializationOptions, VEntityProperty>>(Expression.Block(vars, expressions), new[] { RESULT_FUNC_PARAM_instance, RESULT_FUNC_PARAM_options });
            return lamda.Compile();
        }
```

Calling the function with 
```
    public class DummyClass
    {
        public string Value { get; set; } = "";

        public string? NullableValueSet { get; set; }

        public string? NullableValueNotSet { get; set; }

    }
```
as the input type it generates the following code (Transpiled from the generated debug code string):

```
            Func<DummyClass, VSerializationOptions, VEntityProperty> serializeMember =
            (DummyClass instance, VSerializationOptions options) =>
            {

                VEntityProperty result;
                Dictionary<String, VEntityProperty> subProp;
                
                subProp = ((VSerializeable)instance).Serialize(options);

                result = new VEntityProperty((IDictionary<String, VEntityProperty>)subProp);

                return result;

            };
```

The generated code would be valid if non-nullable reference types where turned off. With non-nullable reference types turned on a compiler error appears:
![image](https://user-images.githubusercontent.com/44103563/80093755-5d327380-8565-11ea-84f3-bc4a90d255b7.png)

It can be fixed by adding a null check:
Generated code:
![image](https://user-images.githubusercontent.com/44103563/80094246-3aed2580-8566-11ea-9778-1b87b4bbcfb8.png)
Code generating code:
```
        public static Func<T, VSerializationOptions, VEntityProperty> GenerateSerializationFunction<T>()
        {
            var WriteResultDictType = typeof(Dictionary<string, VEntityProperty>);
            var SerializeMethod = typeof(VSerializeable).GetMethod(nameof(VSerializeable.Serialize));

            ConstructorInfo entityPropConstructor;
            ParameterExpression entityPropConstructorInput;
            ParameterExpression RESULT_FUNC_PARAM_instance = Expression.Parameter(typeof(T), "instance"); // Input of the result func
            ParameterExpression RESULT_FUNC_PARAM_options = Expression.Parameter(typeof(VSerializationOptions), "options"); // Wheather or not to use a short property name
            ParameterExpression RESULT_FUNC_VAR_result = Expression.Variable(typeof(VEntityProperty), "result");

            var vars = new List<ParameterExpression> { RESULT_FUNC_VAR_result };
            var expressions = new List<Expression>();

            ParameterExpression RESULT_FUNC_VAR_subProp = Expression.Variable(WriteResultDictType, "subProp"); // Used to save sub properties. E. g. when serializing a list or a complex subobject
            vars.Add(RESULT_FUNC_VAR_subProp);

            var ifNotNull = Expression.Assign(RESULT_FUNC_VAR_subProp, Expression.Call(Expression.Convert(RESULT_FUNC_PARAM_instance, typeof(VSerializeable)), SerializeMethod, new[] { RESULT_FUNC_PARAM_options }));

            var ifNull = Expression.Assign(RESULT_FUNC_VAR_subProp, Expression.Convert(Expression.Constant(null), WriteResultDictType));

            // RESULT_FUNC_EXPRESSION Serialize the property
            expressions.Add(Expression.IfThenElse(Expression.Equal(RESULT_FUNC_PARAM_instance, Expression.Constant(null)), ifNull, ifNotNull));

            // Get the constructor that accepts the results of Serialize.
            entityPropConstructor = typeof(VEntityProperty).GetConstructor(new[] { WriteResultDictType })!;
            entityPropConstructorInput = RESULT_FUNC_VAR_subProp;

            // RESULT_FUNC_EXPRESSION Create a new entity property with the serialization result and assign it to entityProp
            expressions.Add(Expression.Assign(RESULT_FUNC_VAR_result, Expression.New(entityPropConstructor, new Expression[] { Expression.Convert(entityPropConstructorInput, entityPropConstructor.GetParameters().First().ParameterType) })));

            // RESULT_FUNC_EXPRESSION return the result
            expressions.Add(RESULT_FUNC_VAR_result);

            var lamda = Expression.Lambda<Func<T, VSerializationOptions, VEntityProperty>>(Expression.Block(vars, expressions), new[] { RESULT_FUNC_PARAM_instance, RESULT_FUNC_PARAM_options });
            return lamda.Compile();
        }
```

Adding the null check solves the problem with Lambda.Compile(), it now returns a valid func.

The issue is that the error is not transported. The only way to see that something is wrong is that the .Compile() function returns null. The code shown here is a simplified version of the actual code generating code I am using. I am using another code generating function which embeds the compiled function shown here as a constant into the one its generating. This makes the issue worse, because now the Top level function now also returns null when compiled, even though the error is produced one level below.

I believe there are two possible solutions to the Problem:
1. Compile() should throw an exception if the Lambda cannot be compiled.
2. The Linq.Expressions system should add support for non-nullable reference types and should throw appropriate runtime exception

**Comment by cston** [comment]
@sidney-pauly 
> Compile() should throw an exception if the Lambda cannot be compiled.

Is the `Expression.Lambda` constructed successfully but `lambda.Compile()` returns `null`?

When trying your repro, I hit several exceptions when I was missing methods or parameters in types such as `VEntityProperty`, but after resolving those, `lambda.Compile()` succeeded.

**Comment by sidney-pauly** [comment]
@cston lamda.Compile() can be run, without throwing an exception, but no valid function is returned. If I remember correctly there where two cases:
- Compiling the Lambda returned null.
- Evaluating the parameter with the Visual Studio 2019 debugger displayed something like "<Internal Error evaluating expression>"

The bug can only be recreated if you enable Non-Nullable-Reference-Types as only then a the compile error gets produced (see the transpiled code).
Maybe there are some other compile settings that also contribute to the error (e. g. Making all warnings errors). Here is the full .csporj:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.1</TargetFramework>
    <Nullable>enable</Nullable>
    <AssemblyName>Core.Utils</AssemblyName>
  </PropertyGroup>

  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
  </ItemGroup>

</Project>
```

If you still cannot reproduce the bug there is the possibility that some other part of the code build on top causes it (although I think that's highly unlikely, as I tested the provided example fairly isolated). If this is the case though I might publish the full code as an Open Source project in the future (This is planned anyways, though not currently high priority)

**Comment by cston** [comment]
@sidney-pauly, from your original description, is the issue that `Expression<TDelegate>.Compile()` compiles successfully (that is, produces a lambda rather than throwing an exception) even though the C# code equivalent to the expression tree would have produced warnings from nullable reference type analysis? 

If so, that would be a significant feature, and we are limiting the changes to System.Linq.Expressions to bugs and regressions (see [README.md](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Linq.Expressions/README.md)).

**Comment by cston** [comment]
Closing since we are not accepting feature contributions to System.Linq.Expressions.
